### PR TITLE
Make BinExport a native exporter extension

### PR DESCRIPTION
**Pull request by cblichmann**
As discussed in #6154, I prepared a work-in-progress PR for integrating BinExport as a native Ghidra exporter extension.

- Add the code from https://github.com/google/binexport/tree/main/java with small changes
- Update all Gradle protobuf dependencies to 3.25.1
- Ship a copy of `binexport2.proto` (authoritative version at https://github.com/google/binexport/blob/main/binexport2.proto)
- Performed a full build using `gradle buildGhidra`

Comments welcome, especially since I'm not a Gradle expert.

### CODE REVIEW

**Comment by ryanmkurtz** [comment]
Was the protobuf upgrade necessary?  If not, it would speed things up if we don't do it as part of this PR.

**Comment by cblichmann** [comment]
No, not necessary. I can revert that part. I included this to prevent a Gradle warning about it not being able to find the 3.17 and 3.21 version of the JARs. 

In order to improve build efficiency (and to prevent Gradle from having to download multiple version of the same dependency), we should nevertheless converge to a single version of `protobuf-java` eventually. But no rush, I can file a separate PR for this.

**Comment by ryanmkurtz** [comment]
Thanks.  Ultimately we are going to want to leverage our existing protobuf compiler and jar that the debugger is using, so we shouldn't have to add much to the Base gradle file.  I will talk with the team to figure out the optimal design and get back to you.

**Comment by ryanmkurtz** [comment]
As a next step, can you have your project depend on `":Framework-AsyncComm"` which will get you your protobuf dependency jar, and use `apply from: "${rootProject.projectDir}/gradle/debugger/hasProtobuf.gradle"` to tie into our `generateProto` task?  Ultimately we wouldn't want this (and Base) to depend on the Debugger, but we (the team) can refactor things later so it makes more sense.  The goal of this would be to not introduce new protobuf specific stuff to our build files.

**Comment by ryanmkurtz** [comment]
Also, what is your intention of including the copyright phrases in the IP headers?  This is not standard practice in the PR's we accept so I wanted to make sure we are all on the same page.

File: Ghidra/Features/Base/src/main/java/ghidra/app/util/exporter/BinExport2Builder.java [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+		entries.sort(Entry.comparingByValue());
+		for (var entry : entries) {
+			builder.addOperandBuilder().addExpressionIndex(entry.getValue());
+		}
+	}
+
+	private void buildMnemonics(Map<String, Integer> mnemonicIndices) {
+		monitor.setIndeterminate(true);
+		monitor.setMessage("Computing mnemonic histogram");
+		var mnemonicHist = new HashMap<String, Integer>();
```
Comment (ryanmkurtz) [inline Ghidra/Features/Base/src/main/java/ghidra/app/util/exporter/BinExport2Builder.java:145]: Our convention is to not use `var` in our code.

Comment (cblichmann) [reply Ghidra/Features/Base/src/main/java/ghidra/app/util/exporter/BinExport2Builder.java]: Done. I hope I didn't break the PR by updating and rebasing different workstations...

**Comment by cblichmann** [comment]
> Also, what is your intention of including the copyright phrases in the IP headers? This is not standard practice in the PR's we accept so I wanted to make sure we are all on the same page.

That's a tricky one. I'm not a copyright lawyer, but from a licensing perspective BinExport and
Ghidra are definitely compatible (Apache 2.0 in both cases).

I think it should be ok to move the copyright line to below the regular IP header if that works
for you? I'll have to check with our licensing folks, but I _think_ we could also treat this
PR as a patch to Ghidra, for which Google is might ok with not retaining the sole copyright 
(again, not a lawyer).

**Comment by cblichmann** [comment]
> As a next step, can you have your project depend on `":Framework-AsyncComm"` which will get you your protobuf dependency jar, and use `apply from: "${rootProject.projectDir}/gradle/debugger/hasProtobuf.gradle"` to tie into our `generateProto` task? Ultimately we wouldn't want this (and Base) to depend on the Debugger, but we (the team) can refactor things later so it makes more sense. The goal of this would be to not introduce new protobuf specific stuff to our build files.

Will do. I hope to get to this early next week.

**Comment by ryanmkurtz** [comment]
> > Also, what is your intention of including the copyright phrases in the IP headers? This is not standard practice in the PR's we accept so I wanted to make sure we are all on the same page.
> 
> That's a tricky one. I'm not a copyright lawyer, but from a licensing perspective BinExport and Ghidra are definitely compatible (Apache 2.0 in both cases).
> 
> I think it should be ok to move the copyright line to below the regular IP header if that works for you? I'll have to check with our licensing folks, but I _think_ we could also treat this PR as a patch to Ghidra, for which Google is might ok with not retaining the sole copyright (again, not a lawyer).

The main thing for us is that team ultimately needs to be in full control over modifying it, deleting it, moving it, etc.  An example reason for this is if we modify an API, we don't want to have to wait on you to submit a PR to fix the compilation errors...that will hold up our release.  Not sure if the copyright will affect that.  Having said that, we don't actually want to maintain it if you are willing to continue to submit PR's for it.  Treating this as a patch to Ghidra would be ideal.  If we can't, i'll have to have some discussions with the team to see how we should proceed.  It might be more appropriate to keep it as an extension so you can maintain full control.

**Comment by cblichmann** [comment]
> > As a next step, can you have your project depend on `":Framework-AsyncComm"` which will get you your protobuf dependency jar, and use `apply from: "${rootProject.projectDir}/gradle/debugger/hasProtobuf.gradle"` to tie into our `generateProto` task? Ultimately we wouldn't want this (and Base) to depend on the Debugger, but we (the team) can refactor things later so it makes more sense. The goal of this would be to not introduce new protobuf specific stuff to our build files.
> 
> Will do. I hope to get to this early next week.

A first go at this was not successful:

```
Circular dependency between the following tasks:
:Base:compileJava
+--- :Base:compileJava (*)
+--- :Framework-AsyncComm:compileJava
|    +--- :Base:compileJava (*)
|    +--- :Framework-AsyncComm:compileJava (*)
|    \--- :ProposedUtils:compileJava
|         +--- :Base:compileJava (*)
|         +--- :Framework-AsyncComm:compileJava (*)
|         \--- :ProposedUtils:compileJava (*)
\--- :ProposedUtils:compileJava (*)

(*) - details omitted (listed previously)
```

**Comment by ryanmkurtz** [comment]
You could make a new module for it in `Ghidra/Features/` for it, rather than sticking in in `Features Base`.  Would make it easier to separate the IP too, if that ends up getting complicated.

**Comment by ryanmkurtz** [comment]
@cblichmann Just wanted to make sure you weren't waiting on anything from me for this.  I think we left it with you checking with your legal people?

**Comment by cblichmann** [comment]
No, not waiting for you. Expect a bit of delay, as I was not able to work on this this week due to other commitments.

**Pull request closed without merging**